### PR TITLE
simplewallet: add wallet-dir flag

### LINKED ISSUE #7674: [Feature suggestion] --wallet-dir flag
From IRC

```
22:11 <clavi> can I set a wallet directory for monero-wallet-cli?
22:11 <clavi> the config file only lets me set a single wallet
22:13 <clavi> I want to be able to run something like "monerocli <walletname>" from any directory
22:13 <clavi> without typing the full paths each time
22:17 <clavi> would be neat to have a "--wallet-dir" option in the CLI config file
```

### PR DESCRIPTION
**Pull request by SNeedlewoods**
submitted by jpk68
https://codeberg.org/jpk68/monero/pulls/30

Closes #7674

Made some suggestions and partially tested the code. If someone wants to do more testing for different combinations of flags that would be welcome.

### CODE REVIEW

**Comment by nahuhh** [comment]
I'm not sure i see the point / use case here.

old:
```
./monero-wallet-cli --wallet-file $HOME/mewallets/mywalletfile
```

new:
```
./monero-wallet-cli --wallet-dir $HOME/mewallets

Enter wallet name: mywalletfile
```

Makes no sense?

edit: i forgot that this can be used in a config file

File: src/simplewallet/simplewallet.h [hunk 1]
```diff
@@ -403,6 +403,7 @@ namespace cryptonote
 
   private:
     std::string m_wallet_file;
+    std::string m_wallet_dir;
```
Comment (selsta) [inline src/simplewallet/simplewallet.h]: Unused

File: src/simplewallet/simplewallet.cpp [hunk 2]
```diff
@@ -612,6 +613,10 @@ namespace
 
     return true;
   }
+  static std::string prefix_wallet_dir(const std::string &filename, const boost::program_options::variables_map &vm)
+  {
+    return (command_line::is_arg_defaulted(vm, arg_wallet_dir) ? "" : command_line::get_arg(vm, arg_wallet_dir) + "/") + filename;
```
Comment (selsta) [inline src/simplewallet/simplewallet.cpp]: Double check if this works with absolute path used for filename and with `--wallet-dir ""`.

File: src/simplewallet/simplewallet.cpp [hunk 3]
```diff
@@ -4428,7 +4437,7 @@ bool simple_wallet::init(const boost::program_options::variables_map& vm)
         auto rc = tools::wallet2::make_from_json(vm, false, m_generate_from_json, password_prompter);
         m_wallet = std::move(rc.first);
         password = rc.second.password();
-        m_wallet_file = m_wallet->path();
+        m_wallet_file = prefix_wallet_dir(m_wallet->path(), vm);
```
Comment (selsta) [inline src/simplewallet/simplewallet.cpp]: The wallet is alredy generated at this point, changing it afterwards is likely wrong.

File: src/simplewallet/simplewallet.cpp
Comment (selsta) [inline src/simplewallet/simplewallet.cpp:3919]: This function uses `wallet_dir_path` to check things but here `wallet_path` again. Is this intentional?

File: src/simplewallet/simplewallet.cpp [hunk 4]
```diff
@@ -169,6 +169,7 @@ namespace
 {
   constexpr std::array<std::string_view, 5> allowed_priority_strings = tools::fee_priority_utilities::fee_priority_strings;
   const auto arg_wallet_file = wallet_args::arg_wallet_file();
+  const command_line::arg_descriptor<std::string> arg_wallet_dir = {"wallet-dir", sw::tr("Directory to look for wallet files"), ""};
```
Comment (selsta) [inline src/simplewallet/simplewallet.cpp]: Description needs update if it also changes the location where files are written.

**Review by selsta — CHANGES_REQUESTED** [review]

File: src/simplewallet/simplewallet.cpp (hunk 2 above)
Comment (selsta) [reply src/simplewallet/simplewallet.cpp]: Also this might need `boost::filesystem::path` instead of hardcoding `"/"`.

File: src/simplewallet/simplewallet.cpp
Comment (SNeedlewoods) [reply src/simplewallet/simplewallet.cpp:3919]: `wallet_dir_path` is the `wallet_dir` + `wallet_path`, the `wallet_dir` also gets applied to `m_wallet_file` at a later point, so this prevents from prefixing the `wallet_dir` twice.

File: src/simplewallet/simplewallet.cpp (hunk 3 above)
Comment (SNeedlewoods) [reply src/simplewallet/simplewallet.cpp]: Right, do you agree we can just ignore `--wallet-dir` when using `--generate-from-json`?
Then maybe we should mention that somewhere, though I don't think [here](https://github.com/monero-project/monero/blob/d509c53ca82749f1f680a5f6f4b1bdf50b3dee93/src/wallet/wallet_args.cpp#L73) is the right place, since that is also used by `monero-wallet-rpc` which handles `--wallet-dir` slightly differently.

Comment (selsta) [reply src/simplewallet/simplewallet.cpp]: I would match the behaviour to `--wallet-file`. If that's also ignored then yes.

**Comment by SNeedlewoods** [comment]
This currently does not work with the following combination (relative path in `wallet_dir` and no wallet file):
`monero-wallet-cli --wallet-dir relativePath/`
then you go in "interactive mode" ([src](https://github.com/monero-project/monero/blob/f9fe258d117c3c43f2ad6828218a3cf4e88a0edf/src/simplewallet/simplewallet.cpp#L3862))  to enter a wallet file name.
To check if the file exists, the method needs to apply the `wallet_dir` flag there, this method sets `m_wallet_file` or `m_generate_new`. The same vars are getting used for the flags `--wallet-file` or `--generate-new-wallet` (see [here](https://github.com/monero-project/monero/pull/10670/changes#diff-25b9648b3da507060201d4ce30539c281676672c1af9a8e4d0e16ec0585a987bR4483)). Which leads to applying `wallet_dir` twice in that case.

A possible solution:
If a `wallet_dir` gets applied in `resolve_wallet_path()` return the absolute path, so if it gets called a second time, it will return early here, instead of applying the `wallet_dir` twice:
```
    if (path.is_absolute()) return filename;
```

**Comment by SNeedlewoods** [comment]
There is still one issue, but that's not introduced by this PR (, AFAICT same issue exists for `monero-wallet-rpc`, `monerod`, etc.).
If providing a path as a flag e.g. `--wallet-file ~/wallet` it gets properly expanded to `/home/user/wallet`, if you provide the exact same path in a config file it doesn't expand and instead uses the `~` literally.

File: src/simplewallet/simplewallet.cpp [hunk 5]
```diff
@@ -612,6 +613,19 @@ namespace
...
+  {
+    std::string wallet_dir = command_line::get_arg(vm, arg_wallet_dir);
+    if (wallet_dir == "") return filename;
+    else if (!boost::filesystem::path(wallet_dir).is_absolute())
+    {
+      fail_msg_writer() << "--wallet-dir " << tr("argument is not an absolute path, ignoring it.");
+      return filename;
+    }
+    boost::filesystem::path path(filename);
+    if (path.is_absolute()) return filename;
```
Comment (selsta) [inline src/simplewallet/simplewallet.cpp]: Move this up to the beginning of the function, otherwise it prints a warning message while the help text claims the parameter is ignored when filename path is absolute.

File: src/simplewallet/simplewallet.cpp [hunk 6]
```diff
@@ -612,6 +613,19 @@ namespace
 
     return true;
   }
+  static std::string resolve_wallet_path(const std::string &filename, const boost::program_options::variables_map &vm)
```
Comment (selsta) [inline src/simplewallet/simplewallet.cpp]: Could you add a couple test cases for all the edge cases?

Comment (selsta) [reply src/simplewallet/simplewallet.cpp]: Tests were run locally, we decided to not add them to the test suite for now.

File: src/simplewallet/simplewallet.cpp [hunk 7]
```diff
@@ -612,6 +613,18 @@ namespace
 
     return true;
   }
+  static std::string resolve_wallet_path(const std::string &filename, const std::string &wallet_dir)
+  {
+    boost::filesystem::path path(filename);
+    if (path.is_absolute()) return filename;
+    if (wallet_dir == "") return filename;
+    else if (!boost::filesystem::path(wallet_dir).is_absolute())
```
Comment (selsta) [inline src/simplewallet/simplewallet.cpp]:
```suggestion
    if (wallet_dir.empty()) return filename;
    if (!boost::filesystem::path(wallet_dir).is_absolute())
```

Small nit but `empty()` is preferred here and the `else` is unnecessary.

File: src/simplewallet/simplewallet.cpp [hunk 8]
```diff
@@ -4169,7 +4186,7 @@ bool simple_wallet::init(const boost::program_options::variables_map& vm)
         return false;
       }
 
-      m_wallet_file=m_generate_from_view_key;
+      m_wallet_file = resolve_wallet_path(m_generate_from_view_key, wallet_dir);
```
Comment (selsta) [inline src/simplewallet/simplewallet.cpp]: `resolve_wallet_path` is called twice in this path which could lead to duplicate warnings. Same for `m_generate_from_keys`.

File: src/simplewallet/simplewallet.cpp [hunk 9]
```diff
@@ -3876,6 +3888,9 @@ bool simple_wallet::ask_wallet_create_if_needed()
         "Specify wallet file name (e.g., MyWallet). If the wallet doesn't exist, it will be created.\n"
         "Wallet file name (or Ctrl-C to quit)")
       );
+
+      wallet_path = resolve_wallet_path(wallet_path, wallet_dir);
```
Comment (selsta) [inline src/simplewallet/simplewallet.cpp]: This should be moved after `cin.eof()`, so that it doesn't run on unexpected input.

**Review by selsta — APPROVED** [review]

**Review by plowsof — APPROVED** [review]
